integration-libs/cdc/core/store/effects/cdc-user-addresses.effect.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * SPDX-FileCopyrightText: 2023 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Injectable } from '@angular/core';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { CdcJsService } from '@spartacus/cdc/root';
+import {
+  Address,
+  GlobalMessageService,
+  GlobalMessageType,
+  UserActions,
+  UserAddressConnector,
+  UserAddressService,
+  UserIdService,
+} from '@spartacus/core';
+import { Observable, of } from 'rxjs';
+import { catchError, map, mergeMap, switchMap, take } from 'rxjs/operators';
+
+@Injectable()
+export class CdcUserAddressesEffects {
+  addressFieldKeys = ['line1', 'line2', 'region.name', 'town', 'postalCode'];
+
+  cdcAddUserAddress$: Observable<UserActions.UserAddressesAction> =
+    createEffect(
+      () =>
+        this.actions$.pipe(
+          ofType(UserActions.ADD_USER_ADDRESS_SUCCESS),
+          mergeMap(() => {
+            return this.updateDefaultAdressInCDC().pipe(
+              map((data: any) => {
+                return new UserActions.LoadUserAddressesSuccess(data);
+              }),
+              catchError((error) => {
+                let errorMessage = error?.errorMessage || ' ';
+                this.messageService.add(
+                  errorMessage,
+                  GlobalMessageType.MSG_TYPE_ERROR
+                );
+                return of(new UserActions.AddUserAddressFail(error));
+              })
+            );
+          })
+        ),
+      { dispatch: false } //prevent dispatching duplicate events in the case of CDC success / failure
+    );
+
+  cdcUpdateUserAddress$: Observable<UserActions.UserAddressesAction> =
+    createEffect(
+      () =>
+        this.actions$.pipe(
+          ofType(
+            UserActions.UPDATE_USER_ADDRESS_SUCCESS,
+            UserActions.LOAD_USER_ADDRESSES /* needed because `updateUserAddress$` dispatches a LOAD_USER_ADDRESSES
+                                             in the scenario where an address is set as default from the address book page */
```

Comment:
```suggestion
```

IMHO should not set an address in CDC every time whenever ANY part of Spartacus application just dispatches `UserActions.LOAD_USER_ADDRESSES` action (likely only trying to load addresses.)

Instead of introducing an inefficient workaround to a problem, I recommend solving it in the roots -> Fix improper design of the core `user-addresses.effect.ts`. IMHO the core effect should _always_ (not conditionally!) dispatch the `UpdateUserSuccessAction` when the OCC update call was successful. I can understand the intentions of the authors of the core effect - they wanted to avoid showing a global message when the change was only marking a new address as default. However I'd like it to be implemented in a way that preserves dispatching `UpadeUserSuccessActions` in any conditions when the update was succesfull.

For reference, here's how the current shape of the core effect:
https://github.com/SAP/spartacus/blob/1ee1c7879fe9ba43a67c37dde19831b326db5482/projects/core/src/user/store/effects/user-addresses.effect.ts#L75-L84
and
https://github.com/SAP/spartacus/blob/1ee1c7879fe9ba43a67c37dde19831b326db5482/projects/core/src/user/store/effects/user-addresses.effect.ts#L142-L152

@SherwinVarghese What do you think about the solution described in this sub-PR proposal https://github.com/SAP/spartacus/pull/17015 ? (note: if we go for it, then unit tests might need being adapted)

Reply:
sure @Platonn,

I am more than happy to address this change. 
However, I am afraid if the PR [17015](https://github.com/SAP/spartacus/pull/17015) will be merged before the 6.0 timeline. We have only a little room at the moment.

Do you think we create an issue for now for CDC, assign it to me to fix later (immediately after 6.0)? By then I believe the changes in the core will be available.

Reply:
Sending new address to CDC every time someone fetches addresses sounds like a bit too aggressive calling CDC backend.

I created a sub-PR to your branch only for easier exaplaining my idea of fix - explaining with code instead of words. That said, it requires manual testing and adding unit tests. Feel free to take it as an inspiration to fix the problem of over-frequent calling CDC backend.

I agree with you, we have little time now. Let's treat the problem as a rearly-happening performance bug and then please schedule fixing it after 6.0.

Reply:
Hi @Platonn,

I just noticed that the code snippet https://github.com/SAP/spartacus/blob/1ee1c7879fe9ba43a67c37dde19831b326db5482/projects/core/src/user/store/effects/user-addresses.effect.ts#L142-L152


actually invokes all addresses while setting one address as default in the Address Book landing page where all addresses are displayed. Do you think we can live with the additional call even when all addresses are rendered in the UI and there is no address change?

Reply:
Thank you for asking your question, it’s valid. I was aware of this flaw of the solution, when I proposed it.
I think we can live with this double-loading of addresses for FF. I consider it a performance bug, which can be fixed after FF, and ideally before the final CF or later in 6.x.

Ideally we should not perform loading addresses in the UI of Addresses List, when we've already reloaded it (during our complex operation of updating address in CDC). That said, I'd like to **avoid** the logic of Addresses List component knowing about CDC effects. And vice versa - I'd like to avoid CDC effects to know about that some component might want to load Addresses in a moment.

I can se 2 possible solutions:
a) Change behavior of Addresses List component: just take addresses data from ngrx store, instead of force-reloading them. However this risks showing stale data, in some edge cases. Although probably this could be solved this or another way, I'm too much fond of that.
b) Use switchMap instead of mergeMap in the core effect of loading addresses. Thanks to that, when 2 independent operations simultaneously want to trigger loading addresses, the first http call will be immediately cancelled, and only the second will be sent. So only one http call would be sent.

Here's where we could change mergeMap to switchMap:
https://github.com/SAP/spartacus/blob/cad6d83929ad1db7b34973230a00790a6e2d0fe6/projects/core/src/user/store/effects/user-addresses.effect.ts#L28

Reply:
sure @Platonn, I favour approach 2 as well. thanks for your inputs and observations.